### PR TITLE
Wad 45 client side validation

### PR DESCRIPTION
**Pull request by martin-gk**
Client and server side validation implemented.

### CODE REVIEW

File: features/projects/add.html [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+    <!-- Bootstrap CSS -->
+    <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bootstrap@4.5.3/dist/css/bootstrap.min.css"
+        integrity="sha384-TX8t27EcRE3e/ihU7zmQxVncDAy5uIKz4rEkgIXeMed4M0jlfIDPvg6uqKI2xXr2" crossorigin="anonymous">
+    <link href="https://fonts.googleapis.com/icon?family=Material+Icons" rel="stylesheet">
+    <link rel="stylesheet" href="/css/styles.css">
+    <!-- favicon -->
+    <link rel="icon" type="image/vnd.microsoft.icon" href="/img/favicon.ico">
+
+    <title>FH JOANNEUM Projekte</title>
+    <meta name="description" content="Overview site of projects of the FH-Joanneum" />
```
Comment (m4rien0) [inline features/projects/add.html]:
```suggestion
    <title>Hinzufügen - FH JOANNEUM Projekte</title>
    <meta name="description" content="Add project page for the projects of the FH-Joanneum" />
```

File: features/projects/add.html [hunk 2]
```diff
@@ -0,0 +1,184 @@
+<!doctype html>
```
Comment (m4rien0) [inline features/projects/add.html:1]: Please save with prettier formatting on (lots of unnecessary empty lines)

File: features/projects/add.html [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+                        <input type="text" class="form-control" id="input-designation"
+                            placeholder="Kostenstelle eintragen" />
+
+                    </div>
+                    <div class="form-group col-md-6">
+                        <label for="input-costplace">Kostenstelle</label>
+                        <input type="text" class="form-control" id="input-costplace"
+                            placeholder="Kostenstelle eintragen" required/>
+                        <div class="invalid-feedback">
+                            Bitte zuständige Kostenstelle eintragen
```
Comment (m4rien0) [inline features/projects/add.html]:
```suggestion
                            Bitte zu verwendende Kostenstelle eintragen
```

File: features/projects/add.html [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+                            placeholder="Beschreibung eintragen" />
+                    </div>
+                    <div class="col-md-6"></div>
+
+                    <div class="form group col-sm-6">
+                        <label for="input-employee">Mitarbeiter</label>
+                        <div id="employee-container">
+                            <input type="text" class="form-control mb-2" id="input-employee" placeholder="Mitarbeiter eintragen" />
+                        </div>
+                        <!-- ask mitch about for label and how to implement for dynamically added fields -->
```
Comment (m4rien0) [inline features/projects/add.html]:
```suggestion

```

File: features/projects/add.html [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+                                <button type="button" id="add-milestone" class="btn btn-primary align-content-between"
+                                        onclick="addMilestoneField();">
+                                    <i class="material-icons ">add</i>
+                                </button>
+                            </td>
+                        </tr>
+                    </table>
+                </div>
+            </div>
+            <button class="btn btn-primary mb-2 mt-1 float-right" type="submit">Submit</button>
```
Comment (m4rien0) [inline features/projects/add.html]:
```suggestion
            <button class="btn btn-primary mb-2 mt-1 float-right" type="submit">Projekt erstellen</button>
```

File: features/projects/add.html [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+
+
+
+
+    </main>
+    <!--<div id="busy-indicator">
+        <div class="spinner-border" role="status">
+            <span class="sr-only">Loading...</span>
+        </div>
+    </div> -->
```
Comment (m4rien0) [inline features/projects/add.html]: Remove if not needed, otherwise uncomment

File: features/projects/projectValidationService.js [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+        .isLength({ max: 50 })
+        .withMessage('Maximale Länge überschritten'),
+    body('customer')
+        .notEmpty()
+        .withMessage('Bitte Namen des Kunden eintragen')
+        .trim()
+        .isLength({ max: 20 }),
+    body('costcenter')
+        .notEmpty()
+        .withMessage('Bitte zuständige Kostenstelle eintragen')
```
Comment (m4rien0) [inline features/projects/projectValidationService.js:25]:
```suggestion
        .withMessage('Bitte zu verwendende Kostenstelle eintragen')
```

File: public/css/styles.css [hunk 8]
```diff
@@ -156,6 +156,7 @@ table tbody tr td:last-of-type {
     min-width: 300px;
 }
 
+<<<<<<< HEAD
```
Comment (m4rien0) [inline public/css/styles.css:159]: unresolved merge conflict markers -> resolve/remove

File: public/css/styles.css [hunk 9]
```diff
@@ -190,3 +191,7 @@ table tbody tr td:last-of-type {
         font-size: 1rem;
     }
 }
+=======
+/* test isi */ 
+
+>>>>>>> 41f1d45c61033f624a6d8097a361095a040eb0ee
```
Comment (m4rien0) [inline public/css/styles.css:197]: remove

File: public/js/add.js [hunk 10]
```diff
@@ -0,0 +1,40 @@
+function addEmployeeField() {
+
+    var employee_container = document.getElementById('employee-container');
+    var new_textfield = document.createElement("input");
+    // <button type="button" id="add-employee" class="btn btn-primary align-content-between" onclick="add_fields();">
```
Comment (m4rien0) [inline public/js/add.js:5]: uncomment if needed, otherwise remove

File: public/js/add.js [hunk 11]
```diff
@@ -0,0 +1,40 @@
+function addEmployeeField() {
```
Comment (m4rien0) [inline public/js/add.js:1]: Whole file: Change to camelcase for variable names

File: public/js/client-side-validation.js [hunk 12]
```diff
@@ -0,0 +1,19 @@
+// Example starter JavaScript for disabling form submissions if there are invalid fields
+console.log("in");
```
Comment (m4rien0) [inline public/js/client-side-validation.js]: Is this needed?

File: public/js/client-side-validation.js [hunk 13]
```diff
@@ -0,0 +1,19 @@
+// Example starter JavaScript for disabling form submissions if there are invalid fields
```
Comment (m4rien0) [inline public/js/client-side-validation.js]: Remove

**Review by m4rien0 — CHANGES_REQUESTED** [review]

File: public/js/add.js (hunk 11 above)
Comment (martin-gk) [reply public/js/add.js:1]: add.js is Isis Task, just pulled for client-side testing

Comment (m4rien0) [reply public/js/add.js:1]: alles klar

**Pull request merged**